Zend/zend_object_handlers.c

```diff
@@ -175,6 +176,160 @@ ZEND_API HashTable *zend_std_get_debug_info(zend_object *object, int *is_temp) /
 }
 /* }}} */
 
+
+static int zend_std_call_op_handler(zend_uchar opcode, zval *result, zval *op1, zval *op2) /* {{{ */
+{
+	zend_class_entry *ce;
+	zend_object *zobj;
+	if(Z_TYPE_P(op1) == IS_OBJECT) {
+		zobj = Z_OBJ_P(op1);
+		ce = Z_OBJCE_P(op1);
+	} else if(Z_TYPE_P(op2) == IS_OBJECT) {
+		zobj = Z_OBJ_P(op2);
+		ce = Z_OBJCE_P(op2);
+	} else {
+		return FAILURE;
+	}
+
+	zend_class_entry *orig_fake_scope = EG(fake_scope);
+	zend_fcall_info fci;
+	zend_fcall_info_cache fcic;
+
+	EG(fake_scope) = NULL;
+
+	zval params[2];
+	/* Unary operator handlers have only one argument */
+	if (op2 == NULL) {
+		fci.param_count = 1;
+		params[0] = *op1;
+		fci.params = params;
+	} else { /* binary op handlers like __add are called with two operands op1, op2 */
+		fci.param_count = 2;
+		params[0] = *op1;
+		params[1] = *op2;
+		fci.params = params;
+	}
+
+	fci.size = sizeof(fci);
+	fci.retval = result;
+	fci.no_separation = 1;
+	//ZVAL_UNDEF(&fci.function_name); /* Unused */
+
+	do
+	{
+		fci.object = zobj;
+
+		/* Determine what handler should be used, based on opcode */
+		switch (opcode)
+		{
+		case ZEND_ADD:
+			fcic.function_handler = ce->__add;
+			ZVAL_STRING(&fci.function_name, ZEND_ADD_FUNC_NAME);
+			break;
+		case ZEND_SUB:
+			fcic.function_handler = ce->__sub;
+			ZVAL_STRING(&fci.function_name, ZEND_SUB_FUNC_NAME);
+			break;
+		case ZEND_MUL:
+			fcic.function_handler = ce->__mul;
+			ZVAL_STRING(&fci.function_name, ZEND_MUL_FUNC_NAME);
+			break;
+		case ZEND_DIV:
+			fcic.function_handler = ce->__div;
+			ZVAL_STRING(&fci.function_name, ZEND_DIV_FUNC_NAME);
+			break;
+		case ZEND_POW:
+			fcic.function_handler = ce->__pow;
+			ZVAL_STRING(&fci.function_name, ZEND_POW_FUNC_NAME);
+			break;
+		case ZEND_MOD:
+			fcic.function_handler = ce->__mod;
+			ZVAL_STRING(&fci.function_name, ZEND_MOD_FUNC_NAME);
+			break;
+		case ZEND_CONCAT:
+			fcic.function_handler = ce->__concat;
+			ZVAL_STRING(&fci.function_name, ZEND_CONCAT_FUNC_NAME);
+			break;
+		case ZEND_SL:
+			fcic.function_handler = ce->__shiftLeft;
+			ZVAL_STRING(&fci.function_name, "__shiftLeft");
+			break;
+		case ZEND_SR:
+			fcic.function_handler = ce->__shiftRight;
+			ZVAL_STRING(&fci.function_name, "__shiftRight");
+			break;
+		case ZEND_BW_OR:
+			fcic.function_handler = ce->__bitwiseOr;
+			ZVAL_STRING(&fci.function_name, "__bitwiseOr");
+			break;
+		case ZEND_BW_AND:
+			fcic.function_handler = ce->__bitwiseAnd;
+			ZVAL_STRING(&fci.function_name, "__bitwiseAnd");
+			break;
+		case ZEND_BW_XOR:
+			fcic.function_handler = ce->__bitwiseXor;
+			ZVAL_STRING(&fci.function_name, "__bitwiseXor");
+			break;
+		case ZEND_BW_NOT:
+			fcic.function_handler = ce->__bitwiseNot;
+			ZVAL_STRING(&fci.function_name, "__bitwiseNot");
+			break;
+		default:
+			return FAILURE;
+			break;
+		}
+
+		/* Check if function exists, check on other operand if possible */
+		if (fcic.function_handler == NULL)
+		{
+			if(zobj == Z_OBJ_P(op1) && op2 != NULL && Z_TYPE_P(op2) == IS_OBJECT) {
```

Comment:
This guard is not sufficient if both operands refer to the same object. Currently, e.g. [concat_001.phpt](https://github.com/php/php-src/blob/4b8171e087a7c53be18aef59ceebfbd15c0eafb9/Zend/tests/concat_001.phpt#L30) gets into an infinite loop. Maybe just:
```suggestion
			if(zobj == Z_OBJ_P(op1) && op2 != NULL && Z_TYPE_P(op2) == IS_OBJECT && Z_OBJ_P(op1) != Z_OBJ_P(op2)) {
```
Maybe the same or a similar change has to be done for the respective guard some lines below.

Reply:
Should be fixed now. Thank you for the hint.